File: api/python/examples/disassembler.py
#!/usr/bin/env python
"""Disassemble instructions at an address of an ELF/PE/Mach-O binary.

Uses LIEF's extended disassembly API to iterate over the instructions
located at the given virtual address of the binary.

Note: only available with the extended version of LIEF.

Example:

    $ python disassembler.py /bin/ls 0x4000
"""

import argparse
import sys
from pathlib import Path

import lief


def disassemble(target: lief.Binary, addr: int) -> None:
    for inst in target.disassemble(addr):
        print(inst)


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    parser.add_argument("file", help="Target binary", type=Path)
    parser.add_argument(
        "address", help="Address to disassemble", type=lambda x: int(x, 0)
    )
    args = parser.parse_args()

    target = lief.parse(args.file)
    if target is None:
        print(f"Can't load: {args.file}", file=sys.stderr)
        return 1
    if isinstance(target, lief.COFF.Binary):
        print("COFF binaries are not supported by the disassembler", file=sys.stderr)
        return 1

    disassemble(target, args.address)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: api/rust/examples/src/bin/disassembler.rs
/// This example shows how to disassemble an ELF/PE/Mach-O binary using LIEF's extended API

use std::process::{self, ExitCode};
use lief::assembly::Instruction;

fn disassemble(target: &dyn lief::generic::Binary, address: u64) -> ExitCode {
    for inst in target.disassemble_address(address) {
        println!("{}", inst.to_string());
    }
    ExitCode::SUCCESS
}

fn main() -> ExitCode {
    if !lief::is_extended() {
        println!("This example requires the extended version of LIEF")
    }

    let args = Vec::from_iter(std::env::args());
    if args.len() != 3 {
        println!("Usage: {} <binary> <address>", args[0]);
        return ExitCode::FAILURE;
    }

    let path = &args[1];
    let addr_str = &args[2];

    let mut addr: u64 = 0;
    if addr_str.starts_with("0x") {
        addr = u64::from_str_radix(addr_str.trim_start_matches("0x"), 16).unwrap();
    } else {
        addr = u64::from_str_radix(&addr_str, 10).unwrap();
    }

    let bin = lief::Binary::parse(&path).expect("Can't parse the binary");
    match bin {
        lief::Binary::ELF(elf) => {
            return disassemble(&elf, addr);
        }

        lief::Binary::PE(pe) => {
            return disassemble(&pe, addr);
        }

        lief::Binary::MachO(fat) => {
            let fit = fat.iter().nth(0).unwrap();
            return disassemble(&fit, addr);
        }
    }
}
